server/utils/device-n-location/device-n-location.utils.ts

```diff
@@ -1,66 +1,50 @@
 import { Context } from "@/server/context";
-import { PrismaClient } from "@prisma/client";
+import { PrismaClient, Prisma } from "@prisma/client";
 import { ITXClientDenyList } from "@prisma/client/runtime/library";
 
-export function getDeviceAndLocationId(
-  ctx: Context,
-  prismaClient: Omit<PrismaClient, ITXClientDenyList> = ctx.prisma
-) {
+export async function getDeviceAndLocationId(ctx: Context) {
   if (!ctx.user) {
     throw new Error("Missing `ctx.user`");
   }
 
-  // TODO: Get ip, userAgent, applicationId...
+  // Look for existing device and location first
+  let deviceAndLocation = await ctx.prisma.deviceAndLocation.findFirst({
+    where: {
+      userId: ctx.user.id,
+      deviceNonce: ctx.session.deviceNonce,
+      ip: ctx.session.ip,
+      userAgent: ctx.session.userAgent
+    },
+    select: {
+      id: true
+    }
+  });
 
-  return prismaClient.deviceAndLocation
-    .upsert({
-      select: {
-        id: true,
-      },
-      where: {
-        userDevice: {
-          userId: ctx.user.id,
-          deviceNonce: ctx.session.deviceNonce,
-          ip: ctx.session.ip,
-          userAgent: ctx.session.userAgent,
-        },
-      },
-      create: {
+  // If not found, create a new one
+  if (!deviceAndLocation) {
+    deviceAndLocation = await ctx.prisma.deviceAndLocation.create({
+      data: {
         deviceNonce: ctx.session.deviceNonce,
         ip: ctx.session.ip,
         userAgent: ctx.session.userAgent,
         userId: ctx.user.id,
-        applicationId: null,
+        createdAt: new Date()
       },
-      update: {},
-    })
-    .then((result) => result.id);
-}
-
-export function getDeviceAndLocationConnectOrCreate(ctx: Context) {
-  if (!ctx.user) {
-    throw new Error("Missing `ctx.user`");
+      select: {
+        id: true
+      }
+    });
   }
+  
+  return deviceAndLocation.id;
+}
 
-  // TODO: Get ip, userAgent, applicationId...
+export async function getDeviceAndLocationConnectOrCreate(ctx: Context) {
```

Comment:
What was the issue with the previous version? While this is fine if there's no other way to make it work, it's probably better to get rid of `getDeviceAndLocationConnectOrCreate()` and simply use `getDeviceAndLocationId()` in the functions that need it, making sure we run as many queries as possible in parallel. With the current implementation, everything is run in series, making procedures that use `getDeviceAndLocationConnectOrCreate()` and/or `getUserConnectOrCreate()` slower.

Reply:
Because we're using a connection pool, i learned that prisma handles multi-operations like these in a way where it sends off the operations to different connections so the jwt claims might get set in one connection and not the other.  I switched this back as now we're setting the jwt claims before any type of db connections

Reply:
Ok, but then, as you ended up doing that in `ensureJwtClaimsSet` for all operations, this would work as it was, using `connectOrCreate` and with the function simply returning a plain object synchronously?